IMG/man/chmap.1
.TH CHMAP 1 "MIDI"
.SH NAME
chmap \- remap MIDI channel commands
.SH SYNOPSIS
.B chmap
[newlist\fB=\fPoldlist] ... <file1 >file2
.SH DESCRIPTION
.I Chmap
reads MIDI command streams
and changes old channel numbers
to new channel numbers.
.PP
In the synopsis, \fIoldlist\fP and \fIoldlist\fP represent lists
of channel numbers where each channel number is
specified by a decimal number in the range 1 to 16.
Channel numbers may be concatenated with "\fB,\fP" and ranges
may be represented by the first and last numbers separated by "\fB\-\fP".
Thus, \fI11\-16\fP is equivalent to \fI11,12,13,14,15,16\fP,
or to \fP11\-13,14\-16\fP.
.PP
If the number of channels in \fInewlist\fP and \fIoldlist\fP are the same,
a one-to-one mapping will result.
As the MIDI commands are read from the input and written on the output
those commands that "belong" to an \fIoldlist\fP channel
will be passed to the output after being remapped to the
corresponding \fInewlist\fP channel.
.PP
If the \fIoldlist\fP contains more than one channel and the \fInewlist\fP
contains only one, a many-to-one mapping will result.
As the MIDI commands are read from the input and written on the output
those commands that "belong" to an \fIoldlist\fP channel
will be passed to the output after being remapped to the
single channel specified in the \fInewlist\fP.
.PP
If the \fInewlist\fP contains more than one channel and the \fIoldlist\fP
contains only one, a one-to-many mapping will result.
As the MIDI commands are read from the input and written on the output
those commands that "belong" to the channel mentioned in the \fIoldlist\fP
will be output once for each channel specified in the \fInewlist\fP.
.PP
Multiple arguments are additive, see examples below.
.PP
If no channel mapping arguments are specified, \fIchmap\fP will print
a count of MIDI messages by channel number and no MIDI output will be
generated.
.SH EXAMPLES
.Cs
chmap 15,16=1,2 <bach >rebach
.Ce
This will copy all MIDI commands from bach that refer to channels
1 or 2, reassigning them to channels 15 and 16 respectively.
All MIDI commands referring to other channels will not be copied.
.Cs
chmap 1=1,2,10\-12 <bach >rebach
chmap 1=1 1=2 1=10\-12 <bach >rebach
.Ce
Both syntaxes will copy MIDI commands that refer to channel 1 unchanged,
and will map commands for channels 2, 10, 11, and 12 to channel 1.
.Cs
chmap 2\-4=1 <bach >rebach
.Ce
This will map MIDI commands that refer to channel 1 to
channel 2 \fIand\fP channel 3 \fIand\fP channel 4.
This can be used to great advantage for ``fattening'' a piece.
.Cs
chmap 1,3=1 2,3=2 <bach >rebach
.Ce
This will copy MIDI commands that refer to channels 1 and 2
straight through, making a copy of each in channel 3.
.PP
Note that status bytes and events pertaining to any channel number not
mentioned in any \fIoldlist\fP will be discarded.
.Cs
chmap <bach
20 events from chan 1
2 events from chan 2
221 events from chan 3
221 events from chan 5
84 events from chan 11
83 events from chan 13
.Ce
.SH SEE ALSO
kmap(1), mecho(1)
\fIRoland MPU-401 Technical Reference Manual\fP.
.SH AUTHOR
Peter Langston, Bell Communications Research (bellcore!psl)


IMG/chmap.c
/*
**	CHMAP -- Channel Map; remap Midi channels
**	psl 10/85, revised 12/85
*/
#include <stdlib.h>
#include	<stdio.h>
#include	<midi.h>

#define	NUMCHAN	MIDI_MAX_CHANS
#define	NUMMAP	NUMCHAN

int chanarg(char *arg, int new[], int *newlenp, int old[], int *oldlenp);
int addmap(int nchan, int ochan);
void syntax(char *a0);
int chanlist(char *list, int chans[NUMCHAN]);

int	map[NUMCHAN][NUMMAP];
int	maplen[NUMCHAN];

main(argc, argv)
char	*argv[];
{
	register int i, chan, stat;
	int new[NUMCHAN], old[NUMCHAN], newlen, oldlen;
	int chin[NUMCHAN], oflg;
	long now;
	MCMD *mp;

	oflg = 0;
	while (--argc > 0) {
	    if (chanarg(argv[argc], new, &newlen, old, &oldlen) == -1)
		syntax(argv[0]);
	    if (newlen == oldlen) {
		for (i = 0; i < newlen; i++)
		    if (addmap(new[i], old[i]) == -1)
			syntax(argv[0]);
	    } else if (newlen == 1) {
		for (i = 0; i < oldlen; i++)
		    if (addmap(new[0], old[i]) == -1)
			syntax(argv[0]);
	    } else if (oldlen == 1) {
		for (i = 0; i < newlen; i++)
		    if (addmap(new[i], old[0]) == -1)
			syntax(argv[0]);
	    } else
		syntax(argv[0]);
	    oflg++;
	}
	for (now = 0L; mp = getmcmd(stdin, now); now = mp->when) {
	    if (mp->len < 2 || mp->len > 4) {
		if (oflg)
		    putmcmd(stdout, mp);
	    } else {
		chan = (mp->cmd[0] & M_CHAN_MASK);
		if (oflg) {
		    stat = (mp->cmd[0] & M_CMD_MASK);
		    for (i = 0; i < maplen[chan]; i++) {
			mp->cmd[0] = stat | map[chan][i];
			putmcmd(stdout, mp);
		    }
		} else
		    chin[chan]++;	
	    }
	}
	if (!oflg) {
	    for (i = 0; i < NUMCHAN; i++)
		if (chin[i])
		    fprintf(stderr, "%d event%s from chan %d\n",
		     chin[i], chin[i] == 1? "" : "s", i + 1);
	}
	exit(0);
}

void syntax(a0)
char	*a0;
{
	fprintf(stderr, "Usage: %s [NEW=OLD] ... <file1 >file2\n", a0);
	fprintf(stderr, "where NEW and OLD are comma separated lists");
	fprintf(stderr, " of channel numbers,\nusing '-' for ranges.");
	fprintf(stderr, "  Channel numbers are decimal numbers");
	fprintf(stderr, " in the range 1-16.\n");
	fprintf(stderr, "NEW and OLD must either contain the same number");
	fprintf(stderr, " of channels (one to one),\n");
	fprintf(stderr, "or NEW must be a single channel (many to one),\n");
	fprintf(stderr, "or OLD must be a single channel (one to many).\n");
	fprintf(stderr, "E.g. %s 1-3=1 4=2 copies channel 1 to", a0);
	fprintf(stderr, " 1, 2, and 3, and channel 2 to 4.\n");
	fprintf(stderr, "Note: only channels mentioned are output.\n");
	fprintf(stderr, "If no channel arguments are given,");
	fprintf(stderr, " print statistics on channel usage.\n");
	exit(2);
}

/*
**	Add ochan->nchan to the map for ochan.
*/

int addmap(nchan, ochan)
{
	if (maplen[ochan] >= NUMMAP) {
	    fprintf(stderr, "Too many destinations for %d\n", ochan);
	    return(-1);
	}
	map[ochan][maplen[ochan]++] = nchan;
	return(0);
}

/*
**	Interpret channel arg of the form list=list
**	where "list" is of the form understood by chanlist().
**	Fill in two ordered lists of channels, "new" and "old",
**	and put their lengths in "newlenp" and "oldlenp".
**	Return 0 for success, -1 for syntax error.
*/
int chanarg(arg, new, newlenp, old, oldlenp)
char	*arg;
int	new[], *newlenp, old[], *oldlenp;
{
	register char *cp;

	for (cp = arg; *cp && *cp != '='; cp++);
	if (*cp != '=')
	    return(-1);
	*cp++ = '\0';
	if ((*newlenp = chanlist(arg, new)) == -1)
	    return(-1);
	if ((*oldlenp = chanlist(cp, old)) == -1)
	    return(-1);
	return(0);
}


/*
**	Interpret channel list of the form: #,#-#,...
**	# is a decimal number in the range 1 to 16, and "3-5" == "3,4,5".
**	Save an ordered list of channels specified in "chans" array.
**	Return the number of channels specified, -1 for error.
*/

int chanlist(list, chans)
char	*list;
int	chans[NUMCHAN];
{
	register char *cp;
	register int i, j, n;

	n = 0;
	for (cp = list; *cp; ) {
	    i = j = atoi(cp) - 1;
	    while ('0' <= *cp && *cp <= '9')
		cp++;
	    if (*cp == '-') {
		j = atoi(++cp) - 1;
		while ('0' <= *cp && *cp <= '9')
		    cp++;
	    }
	    if (i < 0 || i >= NUMCHAN || j < 0 || j >= NUMCHAN)
		return(-1);
	    if (i <= j)
		for (; i <= j; chans[n++] = i++);
	    else
		for (; i >= j; chans[n++] = i--);
	    if (*cp++ != ',')
		break;
	}
	return(n);
}
